odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+            raise ValueError('`dtype` {!r} not supported'.format(dtype))
+
+        if device is None:
+            self.__device = cupy.cuda.get_device_id()
+        else:
+            self.__device = int(device)
+
+        dist = kwargs.pop('dist', None)
+        norm = kwargs.pop('norm', None)
+        inner = kwargs.pop('inner', None)
+        weighting = kwargs.pop('weighting', None)
+        exponent = kwargs.pop('exponent', 2.0)
+
+        # Check validity of option combination (3 or 4 out of 4 must be None)
+        if sum(x is None for x in (dist, norm, inner, weighting)) < 3:
+            raise ValueError('invalid combination of options `weighting`, '
+                             '`dist`, `norm` and `inner`')
+        if any(x is not None for x in (dist, norm, inner)) and exponent != 2.0:
+            raise ValueError('`exponent` cannot be used together with '
+                             '`dist`, `norm` and `inner`')
+
+        # Set the weighting
+        if weighting is not None:
+            if isinstance(weighting, Weighting):
+                if weighting.impl != 'cupy':
+                    raise ValueError("`weighting.impl` must be 'cupy', "
+                                     '`got {!r}'.format(weighting.impl))
+                if weighting.exponent != exponent:
+                    raise ValueError('`weighting.exponent` conflicts with '
+                                     '`exponent`: {} != {}'
+                                     ''.format(weighting.exponent, exponent))
+                self.__weighting = weighting
+            else:
+                self.__weighting = _weighting(weighting, exponent)
+
+            # Check (afterwards) that the weighting input was sane
+            if isinstance(self.weighting, CupyTensorSpaceArrayWeighting):
+                if not np.can_cast(self.weighting.array.dtype, self.dtype):
+                    raise ValueError(
+                        'cannot cast from `weighting` data type {} to '
+                        'the space `dtype` {}'
+                        ''.format(dtype_str(self.weighting.array.dtype),
+                                  dtype_str(self.dtype)))
+                if self.weighting.array.shape != self.shape:
+                    raise ValueError('array-like weights must have same '
+                                     'shape {} as this space, got {}'
+                                     ''.format(self.shape,
+                                               self.weighting.array.shape))
+
+        elif dist is not None:
+            self.__weighting = CupyTensorSpaceCustomDist(dist)
+        elif norm is not None:
+            self.__weighting = CupyTensorSpaceCustomNorm(norm)
+        elif inner is not None:
+            self.__weighting = CupyTensorSpaceCustomInner(inner)
+        else:  # all None -> no weighing
+            self.__weighting = CupyTensorSpaceConstWeighting(1.0, exponent)
+
+    @property
+    def device(self):
```

Comment:
Are we sure we want to make device a space attribute, is it not very similar to order in a sense?

Reply:
Fair question. I think it's more clear-cut with `device` in the sense that keeping data on a device is much easier than keeping the order of an array. Also, I would not like to have a situation where you add two elements of the same space but require copies between devices. Here `space.element()` will make the copy transparent.
Another nice side effect of having `device` as a space property is that you can in fact implement multi-GPU parallelism with product spaces :-).

Reply:
Well, how about a `default_device` construction? Just trying to make sure we do the right thing here.

Reply:
Well for me, device 1 vs. device 2 is not much different from CPU vs GPU. And we decided to make those spaces different, for good reasons I think. 

The biggest reason is that is allows you to  to always know where the data is and not have any unwanted and costly transfers between physically separate memory. 
And the same reason applies IMO for the multi-device case. Transfers between them may be less expensive (I don't know exactly), but I'm sure implicit transfers would still be bad.
Another aspect of this is distribution work load. If the space guarantees its data to be on a specific device, you can do things like
```python
>>> rn_cpu = odl.rn((2, 1000000))
>>> rns_gpu = [odl.rn(1000000, impl='cupy', device=i) for i in (0, 1)]
>>> multi_gpu_space = odl.ProductSpace(*rns_gpu)
>>> chunk_op = ChunkingOperator(rn_cpu, multi_gpu_space)  # hypothetical
>>> busy_ops = [VeryExpensiveOperator(rn_gpu) for rn_gpu in rns_gpu]
>>> busy_op = odl.DiagonalOperator(*busy_ops)
>>> gather_op = GatherOperator(multi_gpu_space, rn_cpu)
>>> dist_gpu_op = gather_op * busy_op * chunk_op
```
I find this version of multi-GPU parallelism really appealing.